tests/test_docs.py

```diff
@@ -0,0 +1,136 @@
+        if rewrite_assertions:
+            loader = AssertionRewritingHook(config=request.config)
+            loader.mark_rewrite(module_name)
+        else:
+            loader = None
+
+        module_path = tmp_work_path / f"{module_name}.py"
+        module_path.write_text(source)
+        spec = importlib.util.spec_from_file_location("__main__", str(module_path), loader=loader)
+        module = importlib.util.module_from_spec(spec)
+        try:
+            spec.loader.exec_module(module)
+        except KeyboardInterrupt:
+            print("KeyboardInterrupt")
+
+    return _import_execute
+
+
+def extract_code_chunks(path: Path, text: str, offset: int):
+    rel_path = path.relative_to(ROOT_DIR)
+    for m_code in re.finditer(r"^```(.*?)$\n(.*?)^```", text, flags=re.M | re.S):
+        prefix = m_code.group(1).lower()
+        if not prefix.startswith(("py", "{.py")):
+            continue
+
+        start_line = offset + text[: m_code.start()].count("\n") + 1
+        code = m_code.group(2)
+        end_line = start_line + code.count("\n") + 1
+        source = "\n" * start_line + code
+        if "test='skip'" in prefix:
+            source = "__skip__"
+        yield pytest.param(f"{path.stem}_{start_line}_{end_line}", source, id=f"{rel_path}:{start_line}-{end_line}")
+
+
+def generate_code_chunks(*directories: str):
+    for d in directories:
+        for path in (ROOT_DIR / d).glob("**/*"):
+            if path.suffix == ".py":
+                code = path.read_text()
+                for m_docstring in re.finditer(r'(^\s*)r?"""$(.*?)\1"""', code, flags=re.M | re.S):
+                    start_line = code[: m_docstring.start()].count("\n")
+                    docstring = dedent(m_docstring.group(2))
+                    yield from extract_code_chunks(path, docstring, start_line)
+            elif path.suffix == ".md":
+                code = path.read_text()
+                yield from extract_code_chunks(path, code, 0)
+
+
+@pytest.fixture
+def mock_dictionary():
+    mock = MagicMock()
+    mock.__getitem__.side_effect = lambda key: mock
+    return mock
+
+
+@pytest.fixture
+def mock_async_endpoint() -> AsyncEndpoint:
+    mock = Mock(spec=AsyncEndpoint)
+    mock.model_endpoint = Mock(spec=ModelEndpoint)
+    mock.model_endpoint.id = "test-endpoint"
+    mock.status = Mock(return_value="READY")
+    return mock
+
+
+@pytest.fixture
+def mock_model_bundle() -> ModelBundle:
+    mock = Mock(spec=ModelBundle)
+    mock.id = "test-bundle"
+    return mock
+
+
+@pytest.fixture
+def mock_batch_job():
+    return {"job_id": "test-batch-job", "status": "SUCCESS"}
+
+
+@pytest.mark.parametrize("module_name,source_code", generate_code_chunks("launch", "docs"))
+def test_docs_examples(
+    module_name,
+    source_code,
+    import_execute,
+    mocker,
+    mock_dictionary,
+    mock_model_bundle,
+    mock_async_endpoint,
+    mock_batch_job,
+):
+    mocker.patch("launch.connection.Connection", MagicMock())
```

Comment:
IIRC, we're getting rid of this `Connection` Nucleus-ism after the client overhaul?

Reply:
😢 unfortunately there's still some routes, like bundle upload, that aren't in our OpenAPI schema... I guess we could add them and then we could remove the `Connection` Nucleus-ism